xlsynth-driver/src/dslx_equiv.rs

```diff
@@ -0,0 +1,523 @@
+        .get_one::<String>("solver")
+        .map(|s| s.parse().unwrap());
+
+    let flatten_aggregates = matches
+        .get_one::<String>("flatten_aggregates")
+        .map(|s| s == "true")
+        .unwrap_or(false);
+    let drop_params: Vec<String> = matches
+        .get_one::<String>("drop_params")
+        .map(|s| s.split(',').map(|x| x.trim().to_string()).collect())
+        .unwrap_or_else(Vec::new);
+
+    let strategy = matches
+        .get_one::<String>("parallelism_strategy")
+        .map(|s| s.parse().unwrap())
+        .unwrap_or(ParallelismStrategy::SingleThreaded);
+
+    let lhs_fixed_implicit_activation = matches
+        .get_one::<String>("lhs_fixed_implicit_activation")
+        .map(|s| s.parse().unwrap())
+        .unwrap_or(false);
+    let rhs_fixed_implicit_activation = matches
+        .get_one::<String>("rhs_fixed_implicit_activation")
+        .map(|s| s.parse().unwrap())
+        .unwrap_or(false);
+
+    let tool_path = config.as_ref().and_then(|c| c.tool_path.as_deref());
+
+    // Build IRs (possibly using external converter) always; if solver is toolchain
+    // we may re-use optimized IRs for temp files.
+    let (lhs_ir_text_raw, lhs_mangled_top) = build_ir_for_dslx(
+        lhs_path,
+        lhs_top.unwrap(),
+        dslx_stdlib_path,
+        dslx_path,
+        enable_warnings,
+        disable_warnings,
+        tool_path,
+        type_inference_v2,
+        opt,
+    );
+    let (rhs_ir_text_raw, rhs_mangled_top) = build_ir_for_dslx(
+        rhs_path,
+        rhs_top.unwrap(),
+        dslx_stdlib_path,
+        dslx_path,
+        enable_warnings,
+        disable_warnings,
+        tool_path,
+        type_inference_v2,
+        opt,
+    );
+
+    // Always optimize before proving equivalence to normalize IR and remove
+    // constructs the g8r parser / equivalence engine may not yet support (e.g.
+    // counted_for).
+    let lhs_ir_text = if opt {
+        lhs_ir_text_raw.clone()
+    } else if let Some(tp) = tool_path {
+        let tmp = tempfile::NamedTempFile::new().unwrap();
+        std::fs::write(tmp.path(), &lhs_ir_text_raw).unwrap();
+        run_opt_main(tmp.path(), Some(&lhs_mangled_top), tp)
+    } else {
+        let pkg = IrPackage::parse_ir(&lhs_ir_text_raw, Some(&lhs_mangled_top)).unwrap();
+        xlsynth::optimize_ir(&pkg, &lhs_mangled_top)
+            .unwrap()
+            .to_string()
+    };
+    let rhs_ir_text = if opt {
+        rhs_ir_text_raw.clone()
+    } else if let Some(tp) = tool_path {
+        let tmp = tempfile::NamedTempFile::new().unwrap();
+        std::fs::write(tmp.path(), &rhs_ir_text_raw).unwrap();
+        run_opt_main(tmp.path(), Some(&rhs_mangled_top), tp)
+    } else {
+        let pkg = IrPackage::parse_ir(&rhs_ir_text_raw, Some(&rhs_mangled_top)).unwrap();
+        xlsynth::optimize_ir(&pkg, &rhs_mangled_top)
+            .unwrap()
+            .to_string()
+    };
+
+    if let Some(solver) = solver_choice {
```

Comment:
I feel like we already have this match statement somewhere (that matches on engine gated by cfg and then calls prove_equiv_internal), can/should we consolidate and just supplement with the printing here?